File: internal/store/zones.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"time"
)

// Zone is a suffix this server is authoritative for. A query for a name
// inside an enabled zone is answered from it or refused by it — it is
// never forwarded, which is what separates a zone from the override list
// it replaced.
type Zone struct {
	ID      int64  `json:"id"`
	Name    string `json:"name"` // apex, lowercase, no trailing dot
	Type    string `json:"type"` // primary | secondary | stub | forwarder | internal
	Enabled bool   `json:"enabled"`

	SOANS      string `json:"soa_ns"`
	SOAMbox    string `json:"soa_mbox"`
	SOASerial  uint32 `json:"soa_serial"`
	SOARefresh uint32 `json:"soa_refresh"`
	SOARetry   uint32 `json:"soa_retry"`
	SOAExpire  uint32 `json:"soa_expire"`
	// SOAMinimum is the negative-cache TTL for NXDOMAINs this zone hands
	// out (RFC 2308), not a floor on positive answers.
	SOAMinimum uint32 `json:"soa_minimum"`
	// SOATTL is the header TTL of the SOA record itself, which is a
	// different thing from SOAMinimum: RFC 2308 §5 makes a negative
	// answer's TTL min(SOAMinimum, SOATTL), so the two have to be
	// separately settable for that minimum to mean anything.
	SOATTL uint32 `json:"soa_ttl"`

	Primaries string `json:"primaries"`
	TSIGKeyID int64  `json:"tsig_key_id"`
	ExpiresAt int64  `json:"expires_at"`
	// RefreshedAt is the last transfer that *succeeded*, unix ms; 0 = never.
	RefreshedAt int64 `json:"refreshed_at"`

	// LastError and LastAttempt are how the most recent transfer attempt
	// went, as opposed to the most recent one that worked. Written only by
	// NoteTransferAttempt — see the 0010 migration for the whole reasoning,
	// and for why there is no stored status beside them.
	//
	// LastError is "" when that attempt succeeded, so a zone that recovered
	// stops reporting one. LastAttempt dates it, successful or not; 0 = never
	// attempted. The pair is only meaningful read together: an error with no
	// date is a claim about now made by an unknown past.
	LastError   string `json:"last_error"`
	LastAttempt int64  `json:"last_attempt"`

	// AllowTransfer is who may pull this zone: a comma-separated list of
	// address, CIDR, or key:<tsig name>, in zones.FormatACL's canonical
	// spelling. Empty means deny, and that is the default.
	AllowTransfer string `json:"allow_transfer"`
	// The outbound twin of LastAttempt/LastError: when a peer last asked for
	// this zone, which peer, and why it was refused if it was. Written only by
	// NoteTransferRequest.
	LastXfrAt    int64  `json:"last_xfr_at"`
	LastXfrPeer  string `json:"last_xfr_peer"`
	LastXfrError string `json:"last_xfr_error"`

	// NotifyTo is who this zone tells when it changes: a comma-separated
	// list of host[:port] [key:<tsig name>], in zones.FormatNotifyTo's
	// canonical spelling. Empty means notify nobody, and that is the
	// default. Applies to a primary and to a secondary — a secondary that
	// re-serves what it pulled has its own downstream secondaries.
	NotifyTo string `json:"notify_to"`

	// ForwardTo is where a `forwarder` zone sends the queries it claims: a
	// comma-separated list of host[:port] in zones.FormatForwardTo's
	// canonical spelling. Empty means the zone names no upstreams, and a
	// forwarder zone with none answers SERVFAIL rather than falling
	// through — it still claims the suffix (§9.11.5).
	//
	// Only `forwarder` uses it. A `stub` names its master in Primaries.
	ForwardTo string `json:"forward_to"`

	CreatedAt  int64 `json:"created_at"`
	ModifiedAt int64 `json:"modified_at"`
}

// ZoneRecord is one RR, named relative to its zone's apex.
type ZoneRecord struct {
	ID     int64  `json:"id"`
	ZoneID int64  `json:"zone_id"`
	Name   string `json:"name"` // '@', 'bifrost', '*', '*.nexus'
	Type   string `json:"type"`
	TTL    uint32 `json:"ttl"`
	// RData is presentation format, rdata portion only.
	RData   string `json:"rdata"`
	Enabled bool   `json:"enabled"`
	Comment string `json:"comment"`
}

// ZoneStore manages authoritative zones and their records.
type ZoneStore interface {
	Zones(ctx context.Context) ([]Zone, error)
	Zone(ctx context.Context, id int64) (Zone, error)
	AddZone(ctx context.Context, z Zone) (int64, error)
	UpdateZone(ctx context.Context, z Zone) error
	DeleteZone(ctx context.Context, id int64) error
	// BumpSerial increments soa_serial in SQL rather than read-modify-write,
	// so two concurrent record edits on the same zone can't land on the
	// same serial.
	BumpSerial(ctx context.Context, zoneID int64) error
	Records(ctx context.Context, zoneID int64) ([]ZoneRecord, error)
	// AllRecords returns every record for every zone, grouped by zone ID, in
	// one query. The resolver rebuilds a whole snapshot on reload, so N+1
	// queries per zone would make reload cost scale with zone count.
	AllRecords(ctx context.Context) (map[int64][]ZoneRecord, error)
	AddRecord(ctx context.Context, r ZoneRecord) (int64, error)
	UpdateRecord(ctx context.Context, r ZoneRecord) error
	DeleteRecord(ctx context.Context, id int64) error
	// ReplaceRecords rewrites a zone in one transaction: the rows named by
	// deleteIDs go, updates are applied in place, adds are inserted, and z
	// replaces the zone row itself — all of it or none of it.
	//
	// It exists because a zone-file import (and, from Milestone D, a zone
	// transfer) is a destructive whole-zone replace, not a sequence of
	// independent edits. Run as separate writes, a storage failure part-way
	// through leaves the zone matching neither its new contents nor its old
	// ones, with no record of where it stopped.
	//
	// The zone row is in the same transaction as the records, and that is
	// the point of passing z at all rather than leaving the caller to follow
	// with UpdateZone. The SOA serial is how every consumer of this zone —
	// a secondary above all — decides whether it already has the current
	// contents. Records that committed while the serial that describes them
	// did not is the one inconsistency nothing downstream can detect: the
	// zone answers with new data under an old serial, and a secondary that
	// has already seen that serial will never ask again.
	//
	// The caller supplies z whole, serial included, because what the new
	// serial should be is a question about DNS (RFC 1982 arithmetic, never
	// going backwards), not about storage.
	//
	// It does not write last_error/last_attempt: those belong to
	// NoteTransferAttempt alone. See its doc comment.
	ReplaceRecords(ctx context.Context, z Zone, deleteIDs []int64, updates, adds []ZoneRecord) error
	// NoteTransferAttempt records how one transfer attempt went: at is when
	// it finished (unix ms) and errText is the transfer's own error, or ""
	// when it succeeded.
	//
	// It is a two-column UPDATE, and that narrowness is the whole design
	// rather than an optimisation. The *whole-row* writes to this table —
	// updateZoneSQL, and ReplaceRecords through it — bind every column from a
	// struct the caller read some time earlier, so a transfer outcome
	// recorded through one of those would silently revert any concurrent edit
	// to enabled, primaries or tsig_key_id: the exact defect the transfer
	// install itself had to be fixed for, reached again through a second
	// door. It is also why neither of them may write these two columns —
	// were last_error in updateZoneSQL, an API PATCH that read the row before
	// a failure and committed after it would erase the failure.
	//
	// It is not the only narrow write here (BumpSerial is another, for a
	// related reason: a serial must be incremented in SQL rather than
	// read-modify-written). What matters is that the column sets are
	// disjoint — this statement owns last_error and last_attempt and touches
	// nothing else, and nothing else touches them.
	NoteTransferAttempt(ctx context.Context, zoneID int64, at int64, errText string) error
	// NoteTransferRequest records one *outbound* transfer request: a peer
	// asking this server for the zone, rather than this server asking a
	// primary. errText is "" when the zone was served and the refusal reason
	// when it was not — a refusal is recorded like a success, which is why
	// this is named for the request rather than for one of its outcomes,
	// exactly as its inbound twin NoteTransferAttempt is.
	//
	// It is a three-column UPDATE for the same reason NoteTransferAttempt is
	// a two-column one — see that method's doc comment — and its column set
	// is disjoint from both other writers of this row: updateZoneSQL (and
	// ReplaceRecords through it) never binds last_xfr_at/last_xfr_peer/
	// last_xfr_error, and this statement never binds anything else.
	NoteTransferRequest(ctx context.Context, zoneID, at int64, peer, errText string) error
}

type zoneStore struct{ s *sqlStore }

const zoneColumns = `id, name, type, enabled, soa_ns, soa_mbox, soa_serial, soa_refresh, soa_retry, soa_expire, soa_minimum, soa_ttl, primaries, tsig_key_id, expires_at, refreshed_at, last_error, last_attempt, allow_transfer, last_xfr_at, last_xfr_peer, last_xfr_error, notify_to, forward_to, created_at, modified_at`

func scanZone(row interface{ Scan(...any) error }, z *Zone) error {
	return row.Scan(&z.ID, &z.Name, &z.Type, &z.Enabled, &z.SOANS, &z.SOAMbox, &z.SOASerial, &z.SOARefresh, &z.SOARetry, &z.SOAExpire, &z.SOAMinimum, &z.SOATTL, &z.Primaries, &z.TSIGKeyID, &z.ExpiresAt, &z.RefreshedAt, &z.LastError, &z.LastAttempt, &z.AllowTransfer, &z.LastXfrAt, &z.LastXfrPeer, &z.LastXfrError, &z.NotifyTo, &z.ForwardTo, &z.CreatedAt, &z.ModifiedAt)
}

func (z *zoneStore) Zones(ctx context.Context) ([]Zone, error) {
	rows, err := z.s.db.QueryContext(ctx, `SELECT `+zoneColumns+` FROM zones ORDER BY id`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	// Non-nil so zero zones marshals to `[]`, not `null` — see
	// clientStore.Groups in sql.go. A brand-new instance starts with none.
	out := []Zone{}
	for rows.Next() {
		var zn Zone
		if err := scanZone(rows, &zn); err != nil {
			return nil, err
		}
		out = append(out, zn)
	}
	return out, rows.Err()
}

func (z *zoneStore) Zone(ctx context.Context, id int64) (Zone, error) {
	var zn Zone
	row := z.s.db.QueryRowContext(ctx, z.s.q(`SELECT `+zoneColumns+` FROM zones WHERE id = ?`), id)
	if err := scanZone(row, &zn); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return Zone{}, ErrNotFound
		}
		return Zone{}, err
	}
	return zn, nil
}

func (z *zoneStore) AddZone(ctx context.Context, zn Zone) (int64, error) {
	// The three last_xfr_* columns are deliberately absent here — they
	// default, and are written only by NoteTransferRequest.
	return z.s.insert(ctx, `INSERT INTO zones (name, type, enabled, soa_ns, soa_mbox, soa_serial, soa_refresh, soa_retry, soa_expire, soa_minimum, soa_ttl, primaries, tsig_key_id, expires_at, refreshed_at, allow_transfer, notify_to, forward_to, created_at, modified_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		zn.Name, zn.Type, zn.Enabled, zn.SOANS, zn.SOAMbox, zn.SOASerial, zn.SOARefresh, zn.SOARetry, zn.SOAExpire, zn.SOAMinimum, zn.SOATTL, zn.Primaries, zn.TSIGKeyID, zn.ExpiresAt, zn.RefreshedAt, zn.AllowTransfer, zn.NotifyTo, zn.ForwardTo, zn.CreatedAt, zn.ModifiedAt)
}

// The four write statements below are each written once and used twice:
// on their own, through the sqlStore helpers that autocommit, and inside
// ReplaceRecords' transaction. Sharing the text is what keeps the two
// paths from drifting — a column added to the hand-write path and missed
// on the import path would be a difference no test of either path alone
// could see, and the zone would silently keep whatever the import didn't
// write.
const (
	updateZoneSQL   = `UPDATE zones SET name = ?, type = ?, enabled = ?, soa_ns = ?, soa_mbox = ?, soa_serial = ?, soa_refresh = ?, soa_retry = ?, soa_expire = ?, soa_minimum = ?, soa_ttl = ?, primaries = ?, tsig_key_id = ?, expires_at = ?, refreshed_at = ?, allow_transfer = ?, notify_to = ?, forward_to = ?, modified_at = ? WHERE id = ?`
	insertRecordSQL = `INSERT INTO zone_records (zone_id, name, type, ttl, rdata, enabled, comment) VALUES (?, ?, ?, ?, ?, ?, ?)`
	// zone_id is deliberately not in the SET list: a record is deleted and
	// recreated to move between zones, not updated in place.
	updateRecordSQL = `UPDATE zone_records SET name = ?, type = ?, ttl = ?, rdata = ?, enabled = ?, comment = ? WHERE id = ?`
	deleteRecordSQL = `DELETE FROM zone_records WHERE id = ?`
)

func updateZoneArgs(zn Zone) []any {
	return []any{zn.Name, zn.Type, zn.Enabled, zn.SOANS, zn.SOAMbox, zn.SOASerial, zn.SOARefresh, zn.SOARetry, zn.SOAExpire, zn.SOAMinimum, zn.SOATTL, zn.Primaries, zn.TSIGKeyID, zn.ExpiresAt, zn.RefreshedAt, zn.AllowTransfer, zn.NotifyTo, zn.ForwardTo, zn.ModifiedAt, zn.ID}
}

func insertRecordArgs(r ZoneRecord) []any {
	return []any{r.ZoneID, r.Name, r.Type, r.TTL, r.RData, r.Enabled, r.Comment}
}

func updateRecordArgs(r ZoneRecord) []any {
	return []any{r.Name, r.Type, r.TTL, r.RData, r.Enabled, r.Comment, r.ID}
}

func (z *zoneStore) UpdateZone(ctx context.Context, zn Zone) error {
	return z.s.execOne(ctx, updateZoneSQL, updateZoneArgs(zn)...)
}

// noteTransferAttemptSQL names the only two columns any transfer outcome is
// allowed to write. See ZoneStore.NoteTransferAttempt for why it is separate
// from updateZoneSQL rather than folded into it.
const noteTransferAttemptSQL = `UPDATE zones SET last_error = ?, last_attempt = ? WHERE id = ?`

func (z *zoneStore) NoteTransferAttempt(ctx context.Context, zoneID int64, at int64, errText string) error {
	return z.s.execOne(ctx, noteTransferAttemptSQL, errText, at, zoneID)
}

// noteTransferRequestSQL names the only three columns an outbound transfer
// outcome may write. Disjoint from updateZoneSQL and from
// noteTransferAttemptSQL — see ZoneStore.NoteTransferRequest.
const noteTransferRequestSQL = `UPDATE zones SET last_xfr_at = ?, last_xfr_peer = ?, last_xfr_error = ? WHERE id = ?`

func (z *zoneStore) NoteTransferRequest(ctx context.Context, zoneID, at int64, peer, errText string) error {
	return z.s.execOne(ctx, noteTransferRequestSQL, at, peer, errText, zoneID)
}

func (z *zoneStore) DeleteZone(ctx context.Context, id int64) error {
	// zone_records rows for this zone go with it — ON DELETE CASCADE in the
	// 0004 migration, not application logic here.
	return z.s.execOne(ctx, `DELETE FROM zones WHERE id = ?`, id)
}

func (z *zoneStore) BumpSerial(ctx context.Context, zoneID int64) error {
	return z.s.execOne(ctx, `UPDATE zones SET soa_serial = soa_serial + 1, modified_at = ? WHERE id = ?`, time.Now().UnixMilli(), zoneID)
}

func (z *zoneStore) Records(ctx context.Context, zoneID int64) ([]ZoneRecord, error) {
	rows, err := z.s.db.QueryContext(ctx, z.s.q(`SELECT id, zone_id, name, type, ttl, rdata, enabled, comment FROM zone_records WHERE zone_id = ? ORDER BY name, id`), zoneID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	// Non-nil so a zone with no records marshals to `[]`, not `null` — see
	// clientStore.Groups in sql.go.
	out := []ZoneRecord{}
	for rows.Next() {
		var r ZoneRecord
		if err := rows.Scan(&r.ID, &r.ZoneID, &r.Name, &r.Type, &r.TTL, &r.RData, &r.Enabled, &r.Comment); err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}

func (z *zoneStore) AllRecords(ctx context.Context) (map[int64][]ZoneRecord, error) {
	rows, err := z.s.db.QueryContext(ctx, `SELECT id, zone_id, name, type, ttl, rdata, enabled, comment FROM zone_records ORDER BY zone_id, name`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[int64][]ZoneRecord{}
	for rows.Next() {
		var r ZoneRecord
		if err := rows.Scan(&r.ID, &r.ZoneID, &r.Name, &r.Type, &r.TTL, &r.RData, &r.Enabled, &r.Comment); err != nil {
			return nil, err
		}
		out[r.ZoneID] = append(out[r.ZoneID], r)
	}
	return out, rows.Err()
}

func (z *zoneStore) AddRecord(ctx context.Context, r ZoneRecord) (int64, error) {
	return z.s.insert(ctx, insertRecordSQL, insertRecordArgs(r)...)
}

func (z *zoneStore) UpdateRecord(ctx context.Context, r ZoneRecord) error {
	return z.s.execOne(ctx, updateRecordSQL, updateRecordArgs(r)...)
}

func (z *zoneStore) DeleteRecord(ctx context.Context, id int64) error {
	return z.s.execOne(ctx, deleteRecordSQL, id)
}

// ReplaceRecords runs the whole replace — records and the zone row — as one
// transaction. See the interface for why the zone row is in it.
//
// The statements are the same ones the single-write methods above issue,
// one per row rather than one batched DELETE ... IN / multi-row INSERT.
// Round trips are not what a replace costs: this is one transaction and one
// commit, where the same work previously cost one commit *per row*, so even
// the largest import a zone file can express (the API caps a file at a
// megabyte, roughly 35,000 records) is cheaper here than it was before.
// Batching would buy the remaining round trips at the price of building
// placeholder lists that differ per dialect and can exceed postgres's
// parameter limit — a real correctness risk for a saving that isn't the
// bottleneck.
func (z *zoneStore) ReplaceRecords(ctx context.Context, zn Zone, deleteIDs []int64, updates, adds []ZoneRecord) error {
	tx, err := z.s.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	// Undoes every statement below unless Commit ran, in which case it is a
	// no-op. Same shape as clientStore.DeleteGroup (crud.go).
	defer tx.Rollback()

	// Deletes, then updates, then adds — the order the API's import applied
	// them in when these were separate writes, kept so the sequence of
	// statements a replace issues is the one already in use rather than a
	// new one whose interaction with a future constraint on zone_records
	// nobody has thought about.
	for _, id := range deleteIDs {
		if err := execOneTx(ctx, tx, z.s.dialect, deleteRecordSQL, id); err != nil {
			return wrapDBErr(err)
		}
	}
	for _, r := range updates {
		if err := execOneTx(ctx, tx, z.s.dialect, updateRecordSQL, updateRecordArgs(r)...); err != nil {
			return wrapDBErr(err)
		}
	}
	for _, r := range adds {
		// Plain Exec rather than the insert helper: a replace has no use for
		// the new row ids, and skipping them is what lets one statement run
		// unchanged on both drivers — postgres needs RETURNING to report an
		// id, sqlite needs LastInsertId.
		if _, err := tx.ExecContext(ctx, z.s.q(insertRecordSQL), insertRecordArgs(r)...); err != nil {
			return wrapDBErr(err)
		}
	}
	// Last, and inside the same transaction: this is the write that carries
	// the new SOA serial, and the failure this method exists to prevent is
	// exactly this statement failing after the records committed.
	if err := execOneTx(ctx, tx, z.s.dialect, updateZoneSQL, updateZoneArgs(zn)...); err != nil {
		return wrapDBErr(err)
	}
	return tx.Commit()
}

File: web/src/lib/zones.ts
import type { Zone } from "../api/types";

/**
 * What a secondary zone's transfer state is, derived from the zone row alone.
 *
 * Every fact used here is a column, and that is the point. The scheduler
 * keeps a richer picture in memory — the consecutive failure count, the
 * pending retry back-off, the error text (zones.Refresher.Status) — but all
 * of that is process-local and dies with the process, so a UI built on it
 * would show a zone that had been failing for a week as healthy for up to a
 * retry interval after every restart.
 *
 * `last_error` and `last_attempt` are the durable half of it (see the 0010
 * migration), written together in one statement by the scheduler and cleared
 * on success. So the screens can state both the state *and* the cause, and
 * both are as current after a restart as before one.
 */

/**
 * The scheduler's floor under every interval taken from an SOA — mirrors
 * `minInterval` in internal/zones/refresh.go. A primary can publish a refresh
 * of 0 (or 1) by accident, and this server clamps rather than obeying it, so
 * computing "next refresh" off the raw SOA value would promise a poll that
 * will not happen.
 */
export const MIN_REFRESH_INTERVAL_MS = 60_000;

/**
 * How often the scheduler asks which zones are due — `refreshTick` in
 * internal/zones/refresh.go. It bounds how *late* an attempt can be, which is
 * why it is the grace period below rather than anything chosen for looks: a
 * zone one second past its deadline is not overdue, it is waiting for the
 * next tick.
 */
export const REFRESH_TICK_MS = 30_000;

/** The interval the scheduler will actually use for this zone, clamped. */
export function refreshIntervalMs(zone: Zone): number {
  return Math.max(zone.soa_refresh * 1000, MIN_REFRESH_INTERVAL_MS);
}

/** The retry interval a failing zone is re-attempted on, clamped the same way. */
export function retryIntervalMs(zone: Zone): number {
  return Math.max(zone.soa_retry * 1000, MIN_REFRESH_INTERVAL_MS);
}

/** When the ordinary schedule next comes round for this zone. */
export function nextRefreshAt(zone: Zone): number {
  return zone.refreshed_at + refreshIntervalMs(zone);
}

/**
 * Why the most recent transfer attempt failed, and when it was made — or null
 * when the most recent attempt succeeded, which is what an empty `last_error`
 * means (the scheduler clears it on success, so a recovered zone stops
 * reporting one).
 *
 * The timestamp comes back with the message on purpose. An error with no date
 * is a claim about the present made by an unknown past: "connection refused"
 * means one thing four minutes after the fact and quite another four days
 * after it, and only the second number tells them apart.
 */
export function lastTransferError(zone: Zone): { message: string; at: number } | null {
  if (zone.last_error === "") return null;
  // The two columns are written together in one statement, so ordinarily an
  // error and the success stamp cannot disagree. The exception is a success
  // whose bookkeeping write failed — best-effort, logged, see
  // Refresher.recordAttempt — which can leave an error older than the last
  // success sitting in the column. Ordering them is what stops that one
  // showing as current.
  if (zone.last_attempt < zone.refreshed_at) return null;
  return { message: zone.last_error, at: zone.last_attempt };
}

/**
 * The part of a transfer error worth reading first.
 *
 * `last_error` is one opaque string — whatever the transfer returned — and for
 * a Go error that is a chain of contexts joined by ": ", outermost first. So
 * the last segment is the innermost failure ("connection refused") and
 * everything in front of it is context the band it appears in already carries:
 * the zone's own name and the primary that was tried.
 *
 * This is a *presentation* split of one string, not a parse of it. Nothing
 * here claims to know what the wording means or to have understood the shape
 * of the message: what comes back is a verbatim suffix, and the caller shows
 * the whole message beside it (a second line on the zone detail page, the
 * `title` on the zones list). So a server that rewords or re-wraps its errors
 * tomorrow costs a nicer first line and nothing else — no text is hidden by
 * getting this wrong, which is the only reason a derivation from unstructured
 * text is defensible at all.
 *
 * When there is no ": " to cut on, or nothing on one side of the last one, the
 * whole message comes back unchanged and there is nothing left to show
 * separately.
 */
export function transferErrorLead(message: string): string {
  const cut = message.lastIndexOf(": ");
  if (cut === -1) return message;
  const head = message.slice(0, cut).trim();
  const tail = message.slice(cut + 2).trim();
  // Both halves have to be real for the cut to be worth making: a message that
  // ends in ": " has no failure after it, and one that starts with it has no
  // context in front of it to set aside.
  if (head === "" || tail === "") return message;
  return tail;
}

/**
 * One of five states, in the order they have to be checked.
 *
 * - `never` — no transfer has ever succeeded. Whether one has been *tried* is
 *   a separate question (`last_attempt`), and the failure that answers it
 *   shows up as the cause beside this state rather than as a state of its
 *   own: either way the zone holds nothing and answers nothing.
 * - `expired` — past the SOA expiry with nothing having succeeded since.
 *   Checked before `failing` because it is the stronger claim: a failing zone
 *   is still serving, an expired one is not.
 * - `failing` — the last attempt failed and said why. Still serving the copy
 *   it has.
 * - `overdue` — no failure recorded, but the refresh deadline came and went
 *   and the copy is still the old one. This is the scheduler not having
 *   reached the zone: it is disabled (skipped entirely), or the process has
 *   only just started. Still serving.
 * - `fresh` — refreshed within the schedule.
 *
 * `never` and `expired` are the two in which the zone answers SERVFAIL rather
 * than serving anything; they are exactly the negation of Zone.Serving in
 * internal/zones/answer.go, and are deliberately derived here by the same two
 * conditions in the same order.
 */
export type TransferState = "never" | "expired" | "failing" | "overdue" | "fresh";

/**
 * **A secondary's states, and only a secondary's.** `expired` is the reason
 * this is not the general answer: `expires_at` is written by a transfer and
 * by nothing else, so on any other type it is 0 — and a row retyped from
 * secondary to stub keeps whatever value it had (see
 * internal/zones/refresh.go, which makes the same carve-out for exactly this
 * reason). A stub is *never given* an expiry, deliberately: its NS set is
 * routing information, and an old-but-working nameserver beats a
 * self-inflicted SERVFAIL. Run one through here and it either dates to the
 * epoch or expires on a stamp that stopped meaning anything the moment it
 * stopped being a secondary.
 *
 * A stub's own state is read off `refreshed_at` and `lastTransferError`
 * alone — neither of which involves an expiry — by `pullState` below, which
 * is what the zone screens ask when the type is not known to be a secondary.
 */
export function transferState(zone: Zone, now: number = Date.now()): TransferState {
  if (zone.refreshed_at === 0) return "never";
  if (zone.expires_at !== 0 && now >= zone.expires_at) return "expired";
  if (lastTransferError(zone) !== null) return "failing";
  // A tick of slack: the scheduler asks which zones are due every 30s, so a
  // zone a moment past its deadline is waiting, not failing. Without this the
  // detail page would flicker into "overdue" once per refresh interval on a
  // perfectly healthy zone.
  if (now >= nextRefreshAt(zone) + REFRESH_TICK_MS) return "overdue";
  return "fresh";
}

/**
 * The two types that pull from a master — the server's own `pullsFromAMaster`
 * (internal/api/zones_handlers.go), which is what decides that `primaries` is
 * required and `tsig_key_id` allowed, and what the scheduler polls.
 *
 * It is also the gate on every health treatment the zone screens give a row.
 * A **forwarder is deliberately outside it**: it claims a suffix and sends
 * live queries on, so its upstreams' reachability is a fact about this
 * instant and nothing about it is stored. There is no column to read, and a
 * list that showed one would be inventing it.
 */
export function pullsFromAMaster(type: Zone["type"]): boolean {
  return type === "secondary" || type === "stub";
}

/**
 * The state of a pulled zone, asked of either type that pulls — and the whole
 * reason it exists is that only one of them may be asked through an expiry.
 *
 * A stub is **never** run through `transferState`. `expires_at` is written by
 * a transfer and by nothing else, and a row retyped from secondary to stub
 * keeps every stamp the transfer left (handleZonePatch sets the type on a row
 * read from the store and clears none of them), so an ordinary stub can carry
 * a long-past expiry that stopped meaning anything the moment it stopped
 * being a secondary. Reading it puts a zone that is routing perfectly well on
 * screen as answering nothing.
 *
 * So a stub gets three states off `refreshed_at` and `lastTransferError`
 * alone, neither of which involves an expiry — the same two the zone detail
 * page's MASTER row derives its note from:
 *
 * - `never` — nothing has ever landed. It holds no NS set, and every name
 *   under the suffix it claims answers SERVFAIL. Not idle: an outage.
 * - `failing` — the last fetch failed; the set it already has is still being
 *   routed on.
 * - `fresh` — the set is current.
 *
 * `expired` is not among them by design (§9.11.8: an old-but-working
 * nameserver beats a self-inflicted SERVFAIL), and neither is `overdue` —
 * that state exists to stop a copy of unknown age reading as current, and a
 * stub's row dates its set outright ("Fetched 5d ago") instead of claiming
 * anything about it.
 *
 * Only meaningful for a type `pullsFromAMaster` accepts; callers gate on it.
 */
export function pullState(zone: Zone, now: number = Date.now()): TransferState {
  if (zone.type === "secondary") return transferState(zone, now);
  if (zone.refreshed_at === 0) return "never";
  return lastTransferError(zone) !== null ? "failing" : "fresh";
}

/**
 * Whether this zone can answer from the records it holds — the client-side
 * twin of Zone.Serving (internal/zones/answer.go). Only a secondary can fail
 * it: a primary owns its data outright.
 *
 * This is what makes "Enabled" an honest badge. A secondary that has never
 * transferred, or whose data has expired, is `enabled` in the database and
 * answering SERVFAIL to every query — so a screen that read `enabled` alone
 * would call it healthy at exactly the moment it is serving nothing.
 */
export function isServing(zone: Zone, now: number = Date.now()): boolean {
  // A stub is deliberately not in here beside the secondary, even though one
  // that has never fetched routes nothing and SERVFAILs its whole suffix. It
  // holds no data on loan, so there is no state it can fall out of; what it
  // has instead is a fetch that has not landed yet, which the MASTER row
  // dates and says outright. Zone.Serving on the server draws the line in the
  // same place.
  if (zone.type !== "secondary") return true;
  const state = transferState(zone, now);
  return state !== "never" && state !== "expired";
}

/**
 * The upstreams a forwarder zone names, one per entry.
 *
 * A count, not a parse: `forward_to` is read back in the server's own
 * canonical spelling (`FormatForwardTo` — port always written, ", "-
 * separated), so splitting it is reading a list the server wrote rather than
 * a second implementation of the grammar. The empty string is no upstreams,
 * which is a configuration and not a gap — see Zone.forward_to.
 */
export function forwardTargets(forwardTo: string): string[] {
  return forwardTo
    .split(",")
    .map((entry) => entry.trim())
    .filter((entry) => entry !== "");
}
